### PR TITLE
[CSL-2492] Run integration tests through TLS connection 

### PR DESCRIPTION
**Pull request by KtorZ**
## Description

I noticed that integration tests are currently running using the HTTP via --wallet-debug option. Integration tests should reproduce the production environment as closely as possible.
The idea here will be to expose smart constructors from Cardano.Wallet.Client.Http to create connection Manager relying on either HTTP or HTTPs.

> **NOTE** /!\ Keep In Sync with #2803 
>
> See https://github.com/input-output-hk/cardano-sl/pull/2803#discussion_r183048392
> Options flags are going to be needed.

## Linked issue

See [CSL-2492](https://iohk.myjetbrains.com/youtrack/issue/CSL-2492)

## Type of change

- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [x] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: wallet-new/integration/CLI.hs [hunk 1]
```diff
@@ -9,28 +9,46 @@ module CLI
...
+    serverPort <- option auto $
+        long        "server-port"
+       <> metavar   "PORT"
+       <> value     8090
+       <> help      "Server port"
+       <> showDefault
+
+    stateless <- switch $
+        long        "stateless"
+        <> help     "If defined - don't persist state between requests."
```
Comment (parsonsmatt) [inline wallet-new/integration/CLI.hs]: Let's delete this option for now, since it's unimplemented.

Comment (KtorZ) [reply wallet-new/integration/CLI.hs]: I wanted to ask about that, good that you bring it on the table ^.^

File: wallet-new/integration/Main.hs [hunk 2]
```diff
@@ -57,8 +62,19 @@ main = do
...
 
-    hspec $ deterministicTests walletClient
+    -- NOTE Our own CLI options interfere with `hspec` which parse them for
+    -- itself when executed, leading to a VERY unclear message:
+    --
+    --     cardano-integration-test: unrecognized option `--tls-ca-cert'
+    --     Try `cardano-integration-test --help' for more information.
+    --
+    -- See also: https://github.com/hspec/hspec/issues/135
+    withArgs [] . hspec $ deterministicTests walletClient
```
Comment (akegalj) [inline wallet-new/integration/Main.hs]: nice

**Comment by KtorZ** [comment]
I am waiting for [DEVOPS-798](https://github.com/input-output-hk/cardano-sl/pull/2803) to be merged, after which I'll make the necessary updates to have it run also as expected on CI :)
If that's okay with you @adinapoli-iohk ^.^

**Comment by parsonsmatt** [comment]
Integration tests are failing:

```haskell
cardano-integration-test: ConnectionError "HttpExceptionRequest Request {
  host                 = \"localhost\"
  port                 = 8090
  secure               = True
  requestHeaders       = []
  path                 = \"/api/v1/wallets\"
  queryString          = \"?page=1&per_page=50\"
  method               = \"GET\"
  proxy                = Nothing
  rawBody              = False
  redirectCount        = 10
  responseTimeout      = ResponseTimeoutDefault
  requestVersion       = HTTP/1.1
}
(ConnectionFailure Network.BSD.getProtocolByName: does not exist (no such protocol name: tcp))"
```

:thinking: That's a weird error. Did you run them locally and they worked?

**Comment by KtorZ** [comment]
I did run them successfully locally and I am as puzzled as you on that one.

**Comment by parsonsmatt** [comment]
Yeah, we might need to get the nix experts in here. It's possible that we're missing some kind of config or package that provides stuff on this.

**Comment by KtorZ** [comment]
Since #2822 was merged, I rebased this one on top of it, using the x509-certificates tool to properly generate server, client and CA certs.

If another pair of eyes could have a look to 1a492776035a5c271dfe9418f7b896967491e1c0, that would be fantastic!

**Comment by parsonsmatt** [comment]
@cleverca22 would you be willing to take a look at the above commit? It looks reasonable to me but my nix-fu is very weak still

**Comment by parsonsmatt** [comment]
I've made PRs to the `servant-quickcheck` library to allow us to pass a manager in explicitly (and fix the MVar problem).

I've backported them to 0.0.4 on my branch: https://github.com/parsonsmatt/servant-quickcheck/compare/v0.0.4...parsonsmatt:matt/backport-mvar-mgr

**Comment by KtorZ** [comment]
Great! Thanks @parsonsmatt!
There's something weird with the commits / changes however right now ^.^ ... I'll clean up a bit the git history we have here.

**Comment by parsonsmatt** [comment]
Alright, we're live :) https://hackage.haskell.org/package/servant-quickcheck-0.0.4.1

**Comment by cleverca22** [comment]
the hydra CI will pass once https://github.com/input-output-hk/iohk-ops/pull/336 has been deployed

File: scripts/launch/demo-cluster/default.nix [hunk 3]
```diff
@@ -103,8 +106,7 @@ in pkgs.writeScript "demo-cluster" ''
   SYNCED=0
   while [[ $SYNCED != 100 ]]
   do
-    # TODO: switch to https when wallet-debug is removed
-    PERC=$(curl --silent -k http://localhost:8090/api/v1/node-info | jq .data.syncProgress.quantity)
+    PERC=$(curl --silent -k --cert ${stateDir}/tls-files/client.pem https://localhost:8090/api/v1/node-info | jq .data.syncProgress.quantity)
```
Comment (disassembler) [inline scripts/launch/demo-cluster/default.nix]: why `-k` here?

Comment (KtorZ) [reply scripts/launch/demo-cluster/default.nix]: Good question. I honestly just added the `--cert` and preserved what was already there. 
Nevertheless, we do have the `ca.crt` however at that stage as well so we could use it all-the-same.

**Pull request merged**